airbyte-integrations/connectors/source-intercom/components.py

```diff
@@ -201,3 +204,130 @@ def migrate(self, stream_state: Mapping[str, Any]) -> Mapping[str, Any]:
         if stream_state.get("companies"):
             migrated_parent_state["companies"] = stream_state.get("companies")
         return {**stream_state, "parent_state": migrated_parent_state}
+
+
+class ResetCursorSignal:
+    """
+    Singleton class that manages a reset signal for Intercom's companies stream.
+    """
+
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+            cls._instance.reset_signal = False
+        return cls._instance
+
+    def is_reset_triggered(self) -> bool:
+        return self.reset_signal
+
+    def trigger_reset(self) -> None:
+        self.reset_signal = True
+
+    def clear_reset(self) -> None:
+        self.reset_signal = False
+
+
+class IntercomErrorHandler(DefaultErrorHandler):
+    """
+    Custom error handler that triggers a reset on HTTP 500 errors.
+    """
+
+    def interpret_response(self, response_or_exception: Optional[Union[requests.Response, Exception]]) -> ErrorResolution:
+        if isinstance(response_or_exception, requests.Response) and response_or_exception.status_code == 500:
+            reset_signal = ResetCursorSignal()
+            reset_signal.trigger_reset()
+            return ErrorResolution(
+                response_action=ResponseAction.RETRY,
+                failure_type=FailureType.transient_error,
+                error_message="HTTP 500 encountered. Triggering reset to retry from the beginning...",
+            )
+        return super().interpret_response(response_or_exception)
+
+
+class IntercomScrollRetriever(SimpleRetriever):
+    """
+    Custom retriever for Intercom's companies stream with reset handling.
```

Comment:
small nit, but let's highlight that this is only compatible with streams that sync using a single date time window instead of multiple windows when the step is defined.

It should be fine for the scroll since it effectively only allows for single-threaded processing but good to highlight this so we don't accidentally use it for future streams.